csrc/multidevice/runtime.h

```diff
@@ -0,0 +1,116 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#ifdef USE_DISTRIBUTED
+#pragma once
+
+#include <disjoint_set.h>
+#include <evaluator_common.h>
+#include <executor.h>
+#include <fusion.h>
+#include <fusion_segmenter.h>
+#include <multidevice/pipeline.h>
+#include <multidevice/pipeline_ir.h>
+#include <scheduler/all_schedulers.h>
+#include <scheduler/registry.h>
+
+#include <ATen/cuda/CUDAContext.h>
+#include <c10/core/DeviceType.h>
+#include <c10/cuda/CUDAFunctions.h>
+#include <multidevice/communicator.h>
+#include <torch/csrc/distributed/c10d/TCPStore.hpp>
+
+#define COMM_BACKEND_DEFAULT CommunicatorBackend::nccl
+#define COMM_SERVER_RANK_DEFAULT 0
+
+namespace nvfuser {
+
+// default canonical mappings rank -> device ID -> concrete devices used in
```

Comment:
do we really need two mappings:
```
rank -> device ID
```
and
```
device ID -> concrete device
```
? Would assuming `rank` to always equal to `device ID` a simplification of the design?

Reply:
Your point is relevant, there are for sure pros and cons.
- On the one hand, since our main target is the case where there is a one-to-one mapping between ranks and device_indices, we could impose the identity function; this would slightly simplify the design.
- On the other hand, maybe we want this mapping not to be the identity, or in the future to be one-to-many or many-to-one. Also, there is a logical difference between device index (abstract index used for scheduling) and the rank (directly retrieves from OMPI environment at runtime)

Since keeping them separated is more general and only moderately complexities the design, I thought we might as well keep them separated. If you disagree I can simply remove it, let me know

Reply:
So "device ID" is the global device ID, for example, if I have 2 nodes, each has 8 GPU, then it would be `0, 1, 2, ..., 15`, but "concrete device" is the device ID local to this current node, so it will always be `0, 1, ..., 7`? What is our model to support multi-node? Will we just have a flattened list of devices from 0 to `num_nodes * gpus_per_node`, or will we also have a node index, and even have a parallel type `nodeIdx.x`?

Reply:
> So "device ID" is the global device ID, for example, if I have 2 nodes, each has 8 GPU, then it would be 0, 1, 2, ..., 15, but "concrete device" is the device ID local to this current node, so it will always be 0, 1, ..., 7? If it is, then would it make sense to rename it to "local device"?

you are right.

Finally, I followed your suggestion and changed the mappings logic to make it simpler and avoid some misleads.
- I kept a mapping `rank <-> dId` but make it way more explicitly the identity. I think this is the right thing for now.
- I also realized that the mapping `dId -> rank` is not really needed. What we only need is to map the current process to a device (i.e., that each rank knows its own device -- a given rank doesn't need to know what is the other rank's device).

This reduces to this now:
```
class TORCH_CUDA_CU_API MultiDeviceRuntime {
...
  // returns the device associated with the process
  auto device() const {
    return at::Device(
        "cuda:" + std::to_string(comm_.local_rank()));
  }

  // returns the rank corresponding to device index
  auto dIdToRank(DeviceIdxType d_id) const {
    return static_cast<RankType>(d_id);
  }

  // returns the device index corresponding to the rank
  auto rankToDiD(RankType rank) const {
    return static_cast<DeviceIdxType>(rank);
  }
...
}
```
You see that this way we can use `local_rank` in `device()`, which I find more natural than `global_rank % at::cuda::getNumGPUs()` as was done previsouly

Reply:
> Also, what is our model to support multi-node? Will we just have a flattened list of devices from 0 to num_nodes * gpus_per_node, or will we also have a node index, and even have a parallel type nodeIdx.x?

This will be the object of future pr. For now this is open. A flattened list of device Idx is enough for the current execution model, but I agree that it probably makes sense to add a "node" parallel type for scheduling

Reply:
This is much clearer to me. Thanks for the change.